services/crosschainconnector/ethereum/service.go

```diff
@@ -2,16 +2,71 @@ package ethereum
 
 import (
 	"context"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/orbs-network/orbs-network-go/instrumentation/log"
+	"github.com/orbs-network/orbs-network-go/services/crosschainconnector/ethereum/adapter"
 	"github.com/orbs-network/orbs-spec/types/go/services"
 )
 
+type ethereumConnectorConfig interface {
+	EthereumEndpoint() string
+}
+
 type service struct {
+	connector adapter.EthereumConnection
+	logger    log.BasicLogger
+	config    ethereumConnectorConfig
 }
 
-func NewEthereumCrosschainConnector() services.CrosschainConnector {
-	return &service{}
+func NewEthereumCrosschainConnector(ctx context.Context, // TODO: why don't we use context here?
+	config ethereumConnectorConfig,
+	connector adapter.EthereumConnection,
+	logger log.BasicLogger) services.CrosschainConnector {
+	s := &service{
+		connector: connector,
+		logger:    logger,
+		config:    config,
+	}
+
+	return s
+}
+
+func (s *service) setupClient() error {
+	if s.connector.GetClient() == nil {
```

Comment:
IMO this entire function should move into the adapter

The service shouldn't bother with restarting clients or connections

The adapter should take in its constructor the endpoint config (it's a config of the adapter, not of the service). Then the adapter will only expose `GetClient()` which will check internally if the client is initialized and initialize it if it isn't.

Reply:
so when we getClient from someplace and the client is nil, the adapter will 'dial' transparently, right?